dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/aggregation/AggregationPipelineQueryNode.java

```diff
@@ -0,0 +1,882 @@
+                varToTripleValue.put(name, SUBJECT);
+                varToTripleHash.put(name, SUBJECT_HASH);
+            }
+            if (sp.getPredicateVar() != null && !sp.getPredicateVar().hasValue()) {
+                String name = sanitize(sp.getPredicateVar().getName());
+                varToTripleValue.put(name, PREDICATE);
+                varToTripleHash.put(name, PREDICATE_HASH);
+            }
+            if (sp.getObjectVar() != null && !sp.getObjectVar().hasValue()) {
+                String name = sanitize(sp.getObjectVar().getName());
+                varToTripleValue.put(name, OBJECT);
+                varToTripleHash.put(name, OBJECT_HASH);
+                varToTripleType.put(name, OBJECT_TYPE);
+            }
+            if (sp.getContextVar() != null && !sp.getContextVar().hasValue()) {
+                String name = sanitize(sp.getContextVar().getName());
+                varToTripleValue.put(name, CONTEXT);
+            }
+        }
+
+        Bson getProjectExpression() {
+            return getProjectExpression(new LinkedList<>(), str -> "$" + str);
+        }
+
+        Bson getProjectExpression(Iterable<String> alsoInclude,
+                Function<String, String> getFieldExpr) {
+            Document values = new Document();
+            Document hashes = new Document();
+            Document types = new Document();
+            for (String varName : varNames()) {
+                values.append(varName, getFieldExpr.apply(valueField(varName)));
+                if (varToTripleHash.containsKey(varName)) {
+                    hashes.append(varName, getFieldExpr.apply(hashField(varName)));
+                }
+                if (varToTripleType.containsKey(varName)) {
+                    types.append(varName, getFieldExpr.apply(typeField(varName)));
+                }
+            }
+            for (String varName : alsoInclude) {
+                values.append(varName, 1);
+                hashes.append(varName, 1);
+                types.append(varName, 1);
+            }
+            List<Bson> fields = new LinkedList<>();
+            fields.add(Projections.excludeId());
+            fields.add(Projections.computed(VALUES, values));
+            fields.add(Projections.computed(HASHES, hashes));
+            if (!types.isEmpty()) {
+                fields.add(Projections.computed(TYPES, types));
+            }
+            fields.add(Projections.computed(LEVEL,
+                    maxValueExpr("$" + LEVEL, getFieldExpr.apply(LEVEL), 0)));
+            fields.add(Projections.computed(TIMESTAMP,
+                    maxValueExpr("$" + TIMESTAMP, getFieldExpr.apply(TIMESTAMP), 0)));
+            return Projections.fields(fields);
+        }
+    }
+
+    /**
+     * Generate a projection expression that evaluates to the maximum of two
+     * fields and a default value.
+     */
+    private static Document maxValueExpr(String field1, String field2, Object defaultValue) {
```

Comment:
why not use mongo's $max?
Add $max to Aggregates?

Reply:
Changed to "$max", weird oversight on my part. Also added a specific check to the appropriate test to convince myself it works.